torch/utils/data/dataloader.py

```diff
@@ -1105,8 +1105,6 @@ def _mark_worker_as_unavailable(self, worker_id, shutdown=False):
         # this method will lead to the worker process ending if _workers_done_event
         # is set, but will only mark it as unavailable for this epoch otherwise.
 
-        assert self._workers_status[worker_id]
```

Comment:
Hmm, when would a worker be marked twice?

Reply:
The worker can be marked twice in the case of `persistent_workers=True`.
First, if the StopIteration has happened, the worker will be marked unavailable, but it won't be finished.

Then, on the shutdown event, the worker will be paused, but we need it to exit the loop, so we send the `None` event even if it was paused.

I will try to improve the checks here and add comments. Maybe the status should not be changed for persistent workers until the real shutdown?

I agree this is not intuitive and clean.

Reply:
First, status should definitely be changed, otherwise the subsequent keys won't be sent to correct workers.

How about this:
1. Let's keep the original invariance: sending `None` means that the worker will be terminated. 
2. So, in the case of `persistent_workers`, and receiving `_IterableDatasetStopIteration`, we simply just mark that worker's `worker_status` to be `False`, *without* calling `_shutdown_worker` (or `_mark_worker_as_unavailable` with the new name). 
3. In worker loop, we also don't do anything special if we receive `None` and `persistent_worker`. 

I personally feel that this way the code will be much simpler, and we won't need to change anything related to `_shutdown_worker`...

Reply:
Let me know what you think!

Reply:
Sounds good!! 
I think my latest commit is in line of that, I will just set explicitly the status to `False` when `_IterableDatasetStopIteration` happens and I think that should be enough.

Reply:
Done, PTAL :)